RecoLocalTracker/SiPhase2VectorHitBuilder/src/VectorHitBuilderAlgorithm.cc

```diff
@@ -0,0 +1,404 @@
+            lparamsLower.second,
+            lparamsUpper.second,
+            pos2Dzx,
+            dir2Dzx,
+            covMat2Dzx,
+            chi22Dzx);
+    LogTrace("VectorHitBuilderAlgorithm") << "\t  pos2Dzx: " << pos2Dzx;
+    LogTrace("VectorHitBuilderAlgorithm") << "\t  dir2Dzx: " << dir2Dzx;
+    LogTrace("VectorHitBuilderAlgorithm") << "\t  cov2Dzx: " << covMat2Dzx;
+    VectorHit2D vh2Dzx = VectorHit2D(pos2Dzx, dir2Dzx, covMat2Dzx, chi22Dzx);
+
+    AlgebraicSymMatrix22 covMat2Dzy;
+    double chi22Dzy = 0.0;
+    Local3DPoint pos2Dzy;
+    Local3DVector dir2Dzy;
+    fit2Dzy(lparamsLower.first,
+            lparamsUpperInLower,
+            lparamsLower.second,
+            lparamsUpper.second,
+            pos2Dzy,
+            dir2Dzy,
+            covMat2Dzy,
+            chi22Dzy);
+    LogTrace("VectorHitBuilderAlgorithm") << "\t  pos2Dzy: " << pos2Dzy;
+    LogTrace("VectorHitBuilderAlgorithm") << "\t  dir2Dzy: " << dir2Dzy;
+    LogTrace("VectorHitBuilderAlgorithm") << "\t  cov2Dzy: " << covMat2Dzy;
+    VectorHit2D vh2Dzy = VectorHit2D(pos2Dzy, dir2Dzy, covMat2Dzy, chi22Dzy);
+
+    OmniClusterRef lowerOmni(lower);
+    OmniClusterRef upperOmni(upper);
+    VectorHit vh = VectorHit(*stack, vh2Dzx, vh2Dzy, lowerOmni, upperOmni);
+    return vh;
+  }
+
+  return VectorHit();
+}
+
+void VectorHitBuilderAlgorithm::fit2Dzx(const Local3DPoint lpCI,
+                                        const Local3DPoint lpCO,
+                                        const LocalError leCI,
+                                        const LocalError leCO,
+                                        Local3DPoint& pos,
+                                        Local3DVector& dir,
+                                        AlgebraicSymMatrix22& covMatrix,
+                                        double& chi2) const {
+  std::vector<float> x = {lpCI.z(), lpCO.z()};
+  std::vector<float> y = {lpCI.x(), lpCO.x()};
+  float sqCI = sqrt(leCI.xx());
+  float sqCO = sqrt(leCO.xx());
+  std::vector<float> sigy = {sqCI, sqCO};
+
+  fit(x, y, sigy, pos, dir, covMatrix, chi2);
+
+  return;
+}
+
+void VectorHitBuilderAlgorithm::fit2Dzy(const Local3DPoint lpCI,
+                                        const Local3DPoint lpCO,
+                                        const LocalError leCI,
+                                        const LocalError leCO,
+                                        Local3DPoint& pos,
+                                        Local3DVector& dir,
+                                        AlgebraicSymMatrix22& covMatrix,
+                                        double& chi2) const {
+  std::vector<float> x = {lpCI.z(), lpCO.z()};
+  std::vector<float> y = {lpCI.y(), lpCO.y()};
+  float sqCI = sqrt(leCI.yy());
+  float sqCO = sqrt(leCO.yy());
+  std::vector<float> sigy = {sqCI, sqCO};
+
+  fit(x, y, sigy, pos, dir, covMatrix, chi2);
+
+  return;
+}
+
+void VectorHitBuilderAlgorithm::fit(const std::vector<float>& x,
```

Comment:
template to accept either array or vector (or even range)

Reply:
or just a sort of SoA (float const *, float const *, float const *, int size)

Reply:
Same here, needs to be vector to be passed to the LinearFitter.

Reply:
Now changed to use arrays thanks to the suggestion by Kevin.